internal/cluster/watch.go

```diff
@@ -0,0 +1,204 @@
+		return
+	}
+
+	found, index, err := indexOfJacocoCoverageFactType(pipelineActivity)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	if found {
+		// Updating the existing jacoco report with the new one. Is this what we want? (HF)
+		pipelineActivity.Spec.Facts[index] = fact
+	} else {
+		pipelineActivity.Spec.Facts = append(pipelineActivity.Spec.Facts, fact)
+	}
+
+	// update the pipeline CRD
+	f = func() error {
+		_, err = jxClient.PipelineActivities(pipelineActivity.Namespace).Update(pipelineActivity)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	err = util.ApplyWithBackoff(f)
+	if err != nil {
+		logger.Errorf("error updating PipelineActivity %s: %s", pipelineActivity.Name, err)
+	} else {
+		logger.Infof("successfully updated PipelineActivity %s with data from %s", pipelineActivity.Name, fact.Original.URL)
+	}
+}
+
+func indexOfJacocoCoverageFactType(pipelineActivity *jenkinsv1.PipelineActivity) (bool, int, error) {
+	found := 0
+	index := -1
+	for i, fact := range pipelineActivity.Spec.Facts {
+		if fact.FactType == jenkinsv1.FactTypeCoverage && util.Contains(fact.Tags, "jacoco") {
+			found++
+			index = i
+		}
+	}
+	if found > 1 {
+		return false, index, errors.Errorf("multiple jacoco facts already attached to PipeLineActivity %s", pipelineActivity.Name)
+	} else if found == 1 {
+		return true, index, nil
+	} else {
+		return false, index, nil
+	}
+}
+
+func createFact(report report.Report, url string) jenkinsv1.Fact {
+	measurements := make([]jenkinsv1.Measurement, 0)
+	for _, c := range report.Counters {
+		t := ""
+		switch c.Type {
+		case "INSTRUCTION":
+			t = jenkinsv1.CodeCoverageCountTypeInstructions
+		case "LINE":
+			t = jenkinsv1.CodeCoverageCountTypeLines
+		case "METHOD":
+			t = jenkinsv1.CodeCoverageCountTypeMethods
+		case "COMPLEXITY":
+			t = jenkinsv1.CodeCoverageCountTypeComplexity
+		case "BRANCH":
+			t = jenkinsv1.CodeCoverageCountTypeBranches
+		case "CLASS":
+			t = jenkinsv1.CodeCoverageCountTypeClasses
+		}
+		measurementCovered := createMeasurement(t, jenkinsv1.CodeCoverageMeasurementCoverage, c.Covered)
+		measurementMissed := createMeasurement(t, jenkinsv1.CodeCoverageMeasurementMissed, c.Missed)
+		measurementTotal := createMeasurement(t, jenkinsv1.CodeCoverageMeasurementTotal, c.Covered+c.Missed)
+		measurements = append(measurements, measurementCovered, measurementMissed, measurementTotal)
+	}
+	fact := jenkinsv1.Fact{
+		FactType: jenkinsv1.FactTypeCoverage,
+		Original: jenkinsv1.Original{
+			URL:      url,
+			MimeType: "application/xml",
+			Tags: []string{
+				"jacoco.xml",
+			},
+		},
+		Tags: []string{
+			"jacoco",
+		},
+		Measurements: measurements,
+		Statements:   []jenkinsv1.Statement{},
+	}
+	return fact
+}
+
+func createMeasurement(t string, measurement string, value int) jenkinsv1.Measurement {
+	return jenkinsv1.Measurement{
+		Name:             fmt.Sprintf("%s-%s", t, measurement),
+		MeasurementType:  "percent",
```

Comment:
See https://github.com/jenkins-x-apps/jx-app-jacoco/issues/95. We can address this in a follow up pull request.